src/rdata.cpp

```diff
@@ -404,8 +406,36 @@ void ReturnData::processBackwardProblem(ForwardProblem const &fwd,
     auto xB = bwd.getAdjointState();
     auto xQB = bwd.getAdjointQuadrature();
 
-    /* NB: This nested loop will not be necessary for fully adjoint
-       preequilibration or post-equilibration without further time points */
+    if (preeq && preeq->getHasQuadrature()) {
+        handleSx0Backward(model, *preeq, llhS0, xQB);
+    } else {
+        handleSx0Forward(model, llhS0, xB, sx_solver);
+    }
+
+    for (int iJ = 0; iJ < model.nJ; iJ++) {
+        for (int ip = 0; ip < model.nplist(); ip++) {
+            if (iJ == 0) {
+                sllh.at(ip) -= llhS0[ip] + xQB[ip * model.nJ];
+            } else {
+                s2llh.at(iJ - 1 + ip * (model.nJ - 1)) -=
+                    llhS0[ip + iJ * model.nplist()] + xQB[iJ + ip * model.nJ];
+            }
+        }
+    }
+}
+
+void ReturnData::handleSx0Backward(Model &model,
```

Comment:
lets learn from past mistakes and add some documentation that gives an intuition why what is done here makes sense.